File: hooks/protect-config.py
#!/usr/bin/env python3
# protect-config.py — PreToolUse:Write/Edit hook
# Blocks modifications to Python project config files.
# Fail-open: parse errors and unexpected exceptions log to stderr and exit 0.
# Exit codes: 0 = allow, 2 = block

import json
import os
import re
import sys

from _harness_dedup import harness_hook_present

PROTECTED_PATTERNS = [
    r"^pyproject\.toml$",
    r"^setup\.cfg$",
    r"^setup\.py$",
    r"^\.ruff\.toml$",
    r"^ruff\.toml$",
    r"^\.flake8$",
    r"^mypy\.ini$",
    r"^\.mypy\.ini$",
    r"^tox\.ini$",
]

# Patterns harness's protect-config.js already guards. When that hook is wired
# we cede these to it and enforce only the Python-unique configs below, so
# ruff.toml edits aren't double-blocked (see #309). Everything not listed here
# (pyproject.toml, setup.cfg, setup.py, .flake8, mypy.ini, tox.ini) is
# Python-specific and has no harness counterpart — canary keeps protecting it.
_HARNESS_COVERED_PATTERNS = {
    r"^\.ruff\.toml$",
    r"^ruff\.toml$",
}


def _effective_patterns():
    if harness_hook_present("protect-config.js"):
        return [p for p in PROTECTED_PATTERNS if p not in _HARNESS_COVERED_PATTERNS]
    return PROTECTED_PATTERNS


def is_protected(file_path, patterns=None):
    base = os.path.basename(file_path)
    patterns = PROTECTED_PATTERNS if patterns is None else patterns
    return any(re.match(p, base) for p in patterns)


def main():
    try:
        raw = sys.stdin.read()
    except Exception:
        sys.stderr.write("[protect-config] Could not read stdin — allowing (fail-open)\n")
        sys.exit(0)

    if not raw.strip():
        sys.stderr.write("[protect-config] Empty stdin — allowing (fail-open)\n")
        sys.exit(0)

    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        sys.stderr.write("[protect-config] Could not parse stdin JSON — allowing (fail-open)\n")
        sys.exit(0)

    try:
        file_path = data.get("tool_input", {}).get("file_path", "")

        if not isinstance(file_path, str) or not file_path:
            sys.stderr.write("[protect-config] Missing file_path in tool input — allowing (fail-open)\n")
            sys.exit(0)

        if is_protected(file_path, _effective_patterns()):
            sys.stderr.write(
                f"BLOCKED: Modification to protected config file: {os.path.basename(file_path)}. "
                "Project config files must not be weakened.\n"
            )
            sys.exit(2)

        sys.exit(0)
    except Exception:
        sys.stderr.write("[protect-config] Unexpected error — allowing (fail-open)\n")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: hooks/protect-config.mjs
#!/usr/bin/env node
// protect-config.mjs — PreToolUse:Write/Edit hook
// Blocks modifications to Python project config files.
// Fail-open: parse errors and unexpected exceptions log to stderr and exit 0.
// Exit codes: 0 = allow, 2 = block
import { basename } from 'node:path';

import { harnessHookPresent } from './_harness_dedup.mjs';

const PROTECTED_PATTERNS = [
  /^pyproject\.toml$/,
  /^setup\.cfg$/,
  /^setup\.py$/,
  /^\.ruff\.toml$/,
  /^ruff\.toml$/,
  /^\.flake8$/,
  /^mypy\.ini$/,
  /^\.mypy\.ini$/,
  /^tox\.ini$/,
];

// Patterns harness's protect-config.js already guards. When that hook is wired
// we cede these to it and enforce only the Python-unique configs, so ruff.toml
// edits aren't double-blocked (see #309).
const HARNESS_COVERED = new Set(['^\\.ruff\\.toml$', '^ruff\\.toml$']);

function effectivePatterns() {
  if (harnessHookPresent('protect-config.js')) {
    return PROTECTED_PATTERNS.filter((p) => !HARNESS_COVERED.has(p.source));
  }
  return PROTECTED_PATTERNS;
}

function isProtected(filePath, patterns = PROTECTED_PATTERNS) {
  const base = basename(filePath);
  return patterns.some((p) => p.test(base));
}

async function readStdin() {
  const chunks = [];
  for await (const chunk of process.stdin) chunks.push(chunk);
  return Buffer.concat(chunks).toString('utf-8');
}

async function main() {
  let raw;
  try {
    raw = await readStdin();
  } catch {
    process.stderr.write(
      '[protect-config] Could not read stdin — allowing (fail-open)\n',
    );
    process.exit(0);
  }
  if (!raw.trim()) {
    process.stderr.write(
      '[protect-config] Empty stdin — allowing (fail-open)\n',
    );
    process.exit(0);
  }

  let data;
  try {
    data = JSON.parse(raw);
  } catch {
    process.stderr.write(
      '[protect-config] Could not parse stdin JSON — allowing (fail-open)\n',
    );
    process.exit(0);
  }

  try {
    const filePath = data?.tool_input?.file_path ?? '';
    if (typeof filePath !== 'string' || !filePath) {
      process.stderr.write(
        '[protect-config] Missing file_path in tool input — allowing (fail-open)\n',
      );
      process.exit(0);
    }

    if (isProtected(filePath, effectivePatterns())) {
      process.stderr.write(
        `BLOCKED: Modification to protected config file: ${basename(filePath)}. ` +
          'Project config files must not be weakened.\n',
      );
      process.exit(2);
    }
    process.exit(0);
  } catch {
    process.stderr.write(
      '[protect-config] Unexpected error — allowing (fail-open)\n',
    );
    process.exit(0);
  }
}

main();
